superset-frontend/src/views/CRUD/utils.tsx

```diff
@@ -380,3 +393,57 @@ export const hasTerminalValidation = (errors: Record<string, any>[]) =>
             isNeedsPassword(payload) || isAlreadyExists(payload),
         ),
   );
+
+function buildDashboardTabsTree(
+  nodeId: string,
+  data: { [key: string]: any },
+  tree: any,
+  traversedIds: any,
+) {
+  if (nodeId in traversedIds) {
+    return traversedIds[nodeId];
+  }
+  const element = data[nodeId];
+
+  const parents: string[] = element.parents.filter(
+    (parent: string) => data[parent] && data[parent].type === 'TAB',
+  );
+  const newNode = {
+    key: nodeId,
+    title: element.meta.text,
+    children: [],
+  };
+
+  // eslint-disable-next-line no-param-reassign
+  traversedIds[nodeId] = newNode;
+  if (parents.length === 0) {
+    tree.push(newNode);
+
+    return newNode;
+  }
+  // Only take the last parent.
+  const parent = parents[parents.length - 1];
+  const parentNode = buildDashboardTabsTree(parent, data, tree, traversedIds);
+  parentNode.children.push(newNode);
+  return newNode;
+}
+
+export function useDashboardTabTree(id: string | number) {
+  const tree: DataNode[] = [];
+  const traversed = {};
+  const dashboardInfo = useDashboard(id);
```

Comment:
i assume this means that the `useDashboardTabTree` always comes with an API call to get the dashboard data. Do you think we should then move this hook into the api hooks file, or at least add a comment for folks who might use this again and not realize it makes an api call?